src/test/dependency-parser.test.ts

```diff
@@ -0,0 +1,358 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {suite} from 'uvu';
+import * as assert from 'uvu/assert';
+import {
+  parseDependency,
+  type ParsedPackage,
+  type ParsedScript,
+} from '../analysis/dependency-parser.js';
+import type {DiagnosticWithoutFile} from '../error.js';
+
+const test = suite<object>();
+
+const cases: Array<
+  [
+    /**
+     * Dependency specifier, e.g. "./pkg#scr".
+     */
+    string,
+
+    /**
+     * A string the same length as the dependency specifier, where each
+     * character represents how we should interpret the corresponding character
+     * from the dependency, with P = package, S = script, E = error, and
+     * anything else is ignored (e.g. "PPPPP_SSS"). This way we can clearly
+     * annotate the expected character ranges for each parsed component.
+     */
+    string,
+    (
+      | {
+          package: ParsedPackage;
+          script: ParsedScript;
+          inverted?: true;
+        }
+      | Omit<DiagnosticWithoutFile, 'location'>
+    ),
+  ]
+> = [
+  [
+    '',
+    '',
+    {
+      severity: 'error',
+      message: 'Dependency cannot be empty',
+    },
+  ],
+  [
+    './pkg#scr',
+    'PPPPP_SSS',
+    {
+      package: {kind: 'path', path: './pkg'},
+      script: {kind: 'name', name: 'scr'},
+    },
+  ],
+  [
+    'scr',
+    'SSS',
+    {
+      package: {kind: 'this'},
+      script: {kind: 'name', name: 'scr'},
+    },
+  ],
+  [
+    '#scr',
+    '_SSS',
+    {
+      severity: 'error',
+      message: 'Package specifier cannot be empty',
+    },
+  ],
+  [
+    './pkg#scr#',
+    'PPPPP_SSSE',
+    {
+      severity: 'error',
+      message: 'Unexpected additional delimiter "#"',
+    },
+  ],
+  [
+    './pkg##',
+    'PPPPP_E',
+    {
+      severity: 'error',
+      message: 'Unexpected additional delimiter "#"',
+    },
+  ],
+  [
+    './pkg',
+    'EEEEE',
+    {
+      severity: 'error',
+      message:
+        `Cross-package dependency must use syntax ` +
+        `"<relative-path>#<script-name>", but there's no ` +
+        `"#" character in "./pkg".`,
+    },
+  ],
+  [
+    String.raw`./\#pkg\##scr`,
+    String.raw`PPPPPPPPP_SSS`,
+    {
+      package: {kind: 'path', path: './#pkg#'},
+      script: {kind: 'name', name: 'scr'},
+    },
+  ],
+  [
+    String.raw`./pkg#\#scr\#`,
+    String.raw`PPPPP_SSSSSSS`,
+    {
+      package: {kind: 'path', path: './pkg'},
+      script: {kind: 'name', name: '#scr#'},
+    },
+  ],
+  [
+    String.raw`\./scr`,
+    String.raw`_SSSSS`,
+    {
+      package: {kind: 'this'},
+      script: {kind: 'name', name: './scr'},
+    },
+  ],
```

Comment:
perhaps an extra test case where leading escaped period `\.` with a `#` or `:` delimiter makes the first segment be interpreted as an npm package?